idx/memory/memory_find_test.go

```diff
@@ -10,89 +10,146 @@ import (
 )
 
 var (
-	ix      idx.MetricIndex
-	queries []query
+	ix         idx.MetricIndex
+	queries    []query
+	tagQueries []tagQuery
 )
 
 type query struct {
 	Pattern         string
 	ExpectedResults int
 }
 
-func cpuMetrics(dcCount, hostCount, hostOffset, cpuCount int, prefix string) []string {
-	series := make([]string, 0)
+type tagQuery struct {
+	Expressions     []string
+	ExpectedResults int
+}
+
+type metric struct {
+	Name string
+	Tags []string
+}
+
+func cpuMetrics(dcCount, hostCount, hostOffset, cpuCount int, prefix string) []metric {
+	var series []metric
 	for dc := 0; dc < dcCount; dc++ {
 		for host := hostOffset; host < hostCount+hostOffset; host++ {
 			for cpu := 0; cpu < cpuCount; cpu++ {
 				p := prefix + ".dc" + strconv.Itoa(dc) + ".host" + strconv.Itoa(host) + ".cpu." + strconv.Itoa(cpu)
-				for _, metric := range []string{"idle", "interrupt", "nice", "softirq", "steal", "system", "user", "wait"} {
-					series = append(series, p+"."+metric)
+				for _, m := range []string{"idle", "interrupt", "nice", "softirq", "steal", "system", "user", "wait"} {
+					series = append(series, metric{
+						Name: p + "." + m,
+						Tags: []string{
+							"dc=dc" + strconv.Itoa(dc),
+							"host=host" + strconv.Itoa(host),
+							"device=cpu",
+							"cpu=cpu" + strconv.Itoa(cpu),
+							"metric=" + m,
+						},
+					})
 				}
 			}
 		}
 	}
 	return series
 }
 
-func diskMetrics(dcCount, hostCount, hostOffset, diskCount int, prefix string) []string {
-	series := make([]string, 0)
+func diskMetrics(dcCount, hostCount, hostOffset, diskCount int, prefix string) []metric {
+	var series []metric
 	for dc := 0; dc < dcCount; dc++ {
 		for host := hostOffset; host < hostCount+hostOffset; host++ {
 			for disk := 0; disk < diskCount; disk++ {
 				p := prefix + ".dc" + strconv.Itoa(dc) + ".host" + strconv.Itoa(host) + ".disk.disk" + strconv.Itoa(disk)
-				for _, metric := range []string{"disk_merged", "disk_octets", "disk_ops", "disk_time"} {
-					series = append(series, p+"."+metric+".read", p+"."+metric+".write")
+				for _, m := range []string{"disk_merged", "disk_octets", "disk_ops", "disk_time"} {
+					series = append(series, metric{
+						Name: p + "." + m + ".read",
+						Tags: []string{
+							"dc=dc" + strconv.Itoa(dc),
+							"host=host" + strconv.Itoa(host),
+							"device=disk",
+							"disk=disk" + strconv.Itoa(disk),
+							"metric=" + m,
+							"direction=read",
+						},
+					})
+					series = append(series, metric{
+						Name: p + "." + m + ".write",
+						Tags: []string{
+							"dc=dc" + strconv.Itoa(dc),
+							"host=host" + strconv.Itoa(host),
+							"device=disk",
+							"disk=disk" + strconv.Itoa(disk),
+							"metric=" + m,
+							"direction=write",
+						},
+					})
 				}
 			}
 		}
 	}
 	return series
 }
 
+func TestMain(m *testing.M) {
+	_tagSupport := tagSupport
+	defer func() { tagSupport = _tagSupport }()
```

Comment:
you can replace both lines with `defer func(t bool) { tagSupport = t}(tagSupport)`

Reply:
is TestMain run for every test in this package? if so, can't we just set tagSupport to true in init() ?

Reply:
We could set it in `Init()`, but then we can't unset it when the test has completed. I'd prefer to not change global settings without changing them back afterwards, because otherwise some other tests/benchmarks might change their behavior depending on whether they run before or after the first one that calls `Init()`, which would make the tests & benchmarks less reliable.

Reply:
k, that makes sense of course.